packages/mdc-chips/chip-set/foundation.js

```diff
@@ -55,69 +56,79 @@ class MDCChipSetFoundation extends MDCFoundation {
     super(Object.assign(MDCChipSetFoundation.defaultAdapter, adapter));
 
     /**
-     * The selected chips in the set. Only used for choice chip set or filter chip set.
-     * @private {!Array<!MDCChipFoundation>}
+     * The ids of the selected chips in the set. Only used for choice chip set or filter chip set.
+     * @private {!Array<string>}
      */
-    this.selectedChips_ = [];
+    this.selectedChipIds_ = [];
   }
 
   /**
-   * Selects the given chip. Deselects all other chips if the chip set is of the choice variant.
-   * @param {!MDCChipFoundation} chipFoundation
+   * Returns an array of the IDs of all selected chips.
+   * @return {!Array<string>}
    */
-  select(chipFoundation) {
-    if (this.adapter_.hasClass(cssClasses.CHOICE)) {
-      this.deselectAll_();
+  getSelectedChipIds() {
+    return this.selectedChipIds_;
+  }
+
+  /**
+   * Toggles selection of the chip with the given id.
+   * @param {string} chipId
+   */
+  toggleSelect(chipId) {
+    if (this.adapter_.hasClass(cssClasses.CHOICE) || this.adapter_.hasClass(cssClasses.FILTER)) {
+      if (this.selectedChipIds_.indexOf(chipId) >= 0) {
+        this.deselect(chipId);
+      } else {
+        this.select(chipId);
+      }
     }
-    chipFoundation.setSelected(true);
-    this.selectedChips_.push(chipFoundation);
   }
 
   /**
-   * Deselects the given chip.
-   * @param {!MDCChipFoundation} chipFoundation
+   * Selects the chip with the given id. Deselects all other chips if the chip set is of the choice variant.
+   * @param {string} chipId
    */
-  deselect(chipFoundation) {
-    const index = this.selectedChips_.indexOf(chipFoundation);
-    if (index >= 0) {
-      this.selectedChips_.splice(index, 1);
+  select(chipId) {
```

Comment:
I'm pretty sure this API creates a bug if called repeatedly for the same choice or filter chip, because it's going to repeatedly push the same chip ID into the array multiple times (without ever clearing the array, even for choice chips).

This won't happen from user interaction since `toggleSelect` will be called which will never call `select` for an already-selected chip, but `select` is also a public API that could be called programmatically.

We should add a test for this bug and fix it. TBH, the earlier behavior seemed simpler - unconditionally deselect everything for choice chips.